core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java

```diff
@@ -0,0 +1,375 @@
+                ShardRouting shard = shardIt.nextOrNull();
+                if (shard != null) {
+                    String nodeId = shard.currentNodeId();
+                    if (!nodeIds.containsKey(nodeId)) {
+                        nodeIds.put(nodeId, Lists.<ShardRouting>newArrayList());
+                    }
+                    nodeIds.get(nodeId).add(shard);
+                } else {
+                    unavailableShardExceptions.add(
+                            new NoShardAvailableActionException(
+                                    shardIt.shardId(),
+                                    "[" + shardIt.shardId().getIndex() + "][" + shardIt.shardId().getId() + "] no shards available while executing " + actionName
+                            )
+                    );
+                }
+            }
+
+            responses = new AtomicReferenceArray<>(nodeIds.size());
+        }
+
+        public void start() {
+            if (nodeIds.size() == 0) {
+                try {
+                    listener.onResponse(newResponse(request, new AtomicReferenceArray(0), unavailableShardExceptions));
+                } catch (Throwable e) {
+                    listener.onFailure(e);
+                }
+            } else {
+                int nodeIndex = -1;
+                for (Map.Entry<String, List<ShardRouting>> entry : nodeIds.entrySet()) {
+                    nodeIndex++;
+                    DiscoveryNode node = nodes.get(entry.getKey());
+                    performOperation(node, entry.getValue(), nodeIndex);
+                }
+            }
+        }
+
+        private void performOperation(final DiscoveryNode node, List<ShardRouting> shards, final int nodeIndex) {
+            try {
+                NodeBroadcastRequest nodeRequest = newNodeRequest(node.getId(), request, shards);
+                transportService.sendRequest(node, transportNodeBroadcastAction, nodeRequest, new BaseTransportResponseHandler<NodeBroadcastResponse>() {
+                    @Override
+                    public NodeBroadcastResponse newInstance() {
+                        return newNodeResponse();
+                    }
+
+                    @Override
+                    public void handleResponse(NodeBroadcastResponse response) {
+                        onOperation(node, nodeIndex, response);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        onFailure(node, nodeIndex, exp);
+                    }
+
+                    @Override
+                    public String executor() {
+                        return ThreadPool.Names.SAME;
+                    }
+                });
+            } catch (Throwable e) {
+                onFailure(node, nodeIndex, e);
+            }
+        }
+
+        protected void onOperation(DiscoveryNode node, int nodeIndex, NodeBroadcastResponse response) {
```

Comment:
onNodeResponse? (operation suggests it's done here, imho, when compare to other on-prefixed namings)